src/core/types.ts

```diff
@@ -51,6 +54,14 @@ export type MutationQueryReducersMap<T = { [key: string]: any }> = {
   [queryName: string]: MutationQueryReducer<T>;
 };
 
+export type MutationUpdaterFunction<T, TVariables, TContext> = (
+  cache: ApolloCache<T>,
+  result: Omit<FetchResult<T>, 'context'>,
```

Comment:
```suggestion
export type MutationUpdaterFunction<
  TData,
  TVariables,
  TContext,
  TCache extends ApolloCache<any>,
> = (
  cache: TCache,
  result: Omit<FetchResult<TData>, 'context'>,
```
@jcreighton As I just recently discovered while working with the old `MutationUpdateFn` type, the `T` type parameter here should probably be called `TData`, which is worth fixing because then it's a little easier to see there's something wrong with `ApolloCache<TData>`. Specifically, the [`ApolloCache`](https://github.com/apollographql/apollo-client/blob/2e2a80a8cb0ef8bbca0eea8d310e869d1a54a5c5/src/cache/core/cache.ts#L14) class is supposed to take a type parameter called `TSerialized`, which describes what you get from `cache.extract()`, and has nothing to do with the shape of individual results (which is what `TData` describes).

Alas, I think this means we need yet another type parameter here, `TCache extends ApolloCache<any>`.

Reply:
Oh, this answers a long-standing "are we sure this is right?" question I had while working with this.